# 2. Extract the company names and details
-------------------
Group 3 , September 24, 2022
1. Gezhi Cheng, 
2. Haowei Lee, 
3. Ziyi Liu, 
4.VS Chaitanya Madduri

> <i>Description: This notebook focuses on analysing the 10-k/10-q files present in the gogole drive and tries to extract the list of companies along with their details </i>





<div class="alert alert-block alert-info">
    <b>Tip:</b> # Please run this notebook in the colab . 
</div> 

### Pre requisites: 
1. And add the shortcut of the drive link : https://drive.google.com/drive/folders/1tZP9A0hrAj8ptNP3VE9weYZ3WDn9jHic to your personal drive.

The above drive contains the 10-k fillings from the author 
Bill McDonald

Do to the huge files we have used our personal google drive folders to save the files.

Files:
Company_Details_analysis.csv - to store the 10-k files of the companies selected

### Output files:
1. And add the shortcut of the drive link :https://drive.google.com/drive/folders/1X4UdGsQiHVWSr63FRiz8rwOuWW5Ua8uI?usp=sharing to your personal drive.

Files:
Selected_10k_v2.csv - to store the 10-k files of the companies selected



## 1. Import Required Packages 

In [2]:

import pandas as pd
import numpy as np
import sklearn 
import os
import zipfile
from sklearn import metrics

In [3]:
# Load the Drive helper and mount 
# Please allow the access to the google drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


## 2.Loading the files to the notebook

### 2.1 Extract the list of the zip files

In [4]:
# getting the list of the zip files in the main folder zip .  
files = [] # empty list to caputure the zip files
dir_path = "/content/drive/MyDrive/10-X_C_Zip" # Main path directory 
  # Iterate directory
for path in os.listdir(dir_path):
    # check if current path is a file
    if os.path.isfile(os.path.join(dir_path, path)):
        files.append(path) # appending the file names to the list

# Printing the number of the files
print(len(files)) 


6


In [5]:
# printing the zip file names 
files

['10-X_C_1993-2000.zip',
 '10-X_C_2001-2005.zip',
 '10-X_C_2006-2010.zip',
 '10-X_C_2011-2015.zip',
 '10-X_C_2016-2020.zip',
 '10-X_C_2021.zip']

In [6]:
# Based on our analysis we have decided to collect all the 10-k fillings between
# 2011-2021

files = [
  '10-X_C_2011-2015.zip',
 '10-X_C_2016-2020.zip',
 '10-X_C_2021.zip']

In [7]:
len(files)

3

In [8]:

DIR_PATH = "/content/drive/MyDrive/SPM_files/" # defining the path in the Google drive

# Profitability Analysis
company_list = pd.read_csv(DIR_PATH + "Company_Details_analysis.csv")

# if you are loading the file from local files
# company_list = pd.read_csv("Company_Details_analysis.csv")

In [9]:
company_list.head()

Unnamed: 0       Company_Name  \
0         206        UNITIL CORP   
1         303  RGC RESOURCES INC   
2         322          SPIRE INC   
3         382      UGI CORP /PA/   
4         437   FIRSTENERGY CORP   

                                 Std_Indust_Class  Index_Key  Class_number  \
0       ELECTRIC & OTHER SERVICES COMBINED [4931]     755001          4931   
1  NATURAL GAS TRANSMISSION & DISTRIBUTION [4923]    1069533          4923   
2                 NATURAL GAS DISTRIBUTION [4924]    1126956          4924   
3            GAS & OTHER SERVICES COMBINED [4932]     884614          4932   
4                        ELECTRIC SERVICES [4911]    1031296          4911   

   Class_number_first  
0                  49  
1                  49  
2                  49  
3                  49  
4                  49

## 3. Creating nexted dictionaries for file location
- we wil use this dictionary to get location of all the 10-k filling we requre from the company list.
- Level 1 will store the main zip files eg : '10-X_C_2011-2015.zip'
- Level 2 wil store the company index numbers eg: 755001
- Level 3 will store the folder paths eg: '2011/QTR1/20110203_10-K_edgar_data_755001_0001193125-11-022690.txt'

In [10]:
final_dic = {}

In [11]:
# Extract the file names and fodler details for the selected companies
# please create the company folders in the destination folder.
drive_path = "/content/drive/MyDrive/10-X_C_Zip/"


for file_name in files:
  final_dic[file_name] = {} 
  for index in company_list['Index_Key'].unique():
    final_dic[file_name][index]= {}
  zip_filepath = drive_path + file_name
  zip = zipfile.ZipFile(drive_path + file_name)
  for index in company_list['Index_Key'].unique():
    final_dic[file_name][index]= [s for s in zip.namelist() if ('_' + str(index)+ '_' in s) & ('10-K_edgar' in s)]


In [12]:
final_dic.keys()

dict_keys(['10-X_C_2011-2015.zip', '10-X_C_2016-2020.zip', '10-X_C_2021.zip'])

In [13]:
# Dictionary contains the location of all the files taht need to be extrated
final_dic['10-X_C_2011-2015.zip'].keys()

dict_keys([755001, 1069533, 1126956, 884614, 1031296, 1610418, 1109357, 1336047, 1623925, 100122, 1317630, 1318220, 73088, 823768, 1582244, 1060391, 1130464, 922224, 753308, 1428389, 1410636, 1522727, 1591763, 1047862, 92122, 1326160, 1379661, 4904, 352541, 1123852, 65984, 764622, 72903, 76063, 81018, 1085392, 92521, 1081316, 1585790, 1693317, 788784, 822818, 1013871, 1021635, 1032208, 1035201, 91928, 1130310, 107833, 1193311, 1423902, 1419242, 1540947, 31978, 1161728, 66004, 72741, 827052, 1002910, 1061219, 715957, 787250, 354707, 78128, 861878, 874761, 91882, 1108426, 1296445, 1561660, 1567683, 1634997, 1637757, 1711269, 1783400, 1733998, 1612720, 1422841, 1549922, 1378140, 1368265, 1061069, 727273, 1043186, 1528356, 1563922, 1599947, 1690384, 891532, 1730346, 1338065, 1004980])

In [14]:
final_dic.keys() # printing main dictionary keys

dict_keys(['10-X_C_2011-2015.zip', '10-X_C_2016-2020.zip', '10-X_C_2021.zip'])

In [ ]:
# Printing an example company key
final_dic['10-X_C_2016-2020.zip'][884614]

['2016/QTR4/20161122_10-K_edgar_data_884614_0000884614-16-000067.txt',
 '2017/QTR4/20171121_10-K_edgar_data_884614_0000884614-17-000029.txt',
 '2018/QTR4/20181120_10-K_edgar_data_884614_0000884614-18-000033.txt',
 '2019/QTR4/20191126_10-K_edgar_data_884614_0000884614-19-000059.txt',
 '2020/QTR4/20201120_10-K_edgar_data_884614_0000884614-20-000052.txt']

## 3. Extract the text data , company name  and details form the zip files

3.1 Extract text content to Dataframe

In [ ]:
# A master dictionary to save the company details
# which details of the company are specified in th elist
columns = ['Company_Key', 'Text_data', 'Quarter_details',"Year"] 
# creating the empty dataframe with columns
master_dataframe = pd.DataFrame(columns= columns) 

In [ ]:

# please create the company folders in the destination folder.
drive_path = "/content/drive/MyDrive/10-X_C_Zip/" # the input folder

#looping around the zip files 
for file_name in files: 
  # zip file handler  
  zip = zipfile.ZipFile(drive_path + file_name)
  # looping around the company index codes that we selected
  for company in company_list['Index_Key'].unique():
    # Checking if the path exists
    if final_dic[file_name][company]:
      # looping around all the file paths 
      for text_file_name in final_dic[file_name][company]: 
         # getting latest index of file to create the next index for entry
        index_val = master_dataframe.shape[0]
        try:
          # capturing the 10-k filling data and filepath
          master_dataframe.loc[index_val,"Text_data"] =  zip.read(text_file_name)
          master_dataframe.loc[index_val,"filename"] =  text_file_name
        except:
          # except block in case if there are errors in the code
          print("Name of the file :: {}".format(text_file_name))
    

In [ ]:
final_dic.keys()

dict_keys(['10-X_C_2011-2015.zip', '10-X_C_2016-2020.zip', '10-X_C_2021.zip'])

## 4.Extracting company details 

In [ ]:
def extract_compnay_name (text_input):
   '''
   This function will extarct the name of the company from text
   input: 
    text_input: string which contains all the 10-k info
   return:
    name of the company
   '''
   file_data = str(text_input).split('\\n')
   for line in file_data:
     stripped_line = line.strip()
     if "COMPANY CONFORMED NAME:" in stripped_line :                
       splitted_str = stripped_line.split(":")  # split COMPANY CONFORMED NAME and real name
       company_name = splitted_str[1] 
       return company_name.strip() 

def extract_company_index (text_input):
   '''
    This function will extarct the name of the company from text
   input: 
    text_input: string which contains all the 10-k info
   return:
    company Index
   '''   
   file_data = str(text_input).split('\\n')
   for line in file_data:
     stripped_line = line.strip()
     if "CENTRAL INDEX KEY:" in stripped_line:
            
      splitted_str = stripped_line.split(":")  # split COMPANY CONFORMED NAME and real name
      company_index = splitted_str[1]
      return company_index.strip() 
      

### 4.1 Extract the Company name.

In [ ]:
# Extract the company name
master_dataframe['company_name'] = master_dataframe['Text_data'].apply(lambda x: extract_compnay_name(x)) 

In [ ]:
# code to remove special charaters from the name 
master_dataframe['company_name'] = master_dataframe['company_name'].str[6:-2]

### 4.2 Extract the 10-k filling year and quarter

In [ ]:
# Extract the Year details from the file path 
master_dataframe['Year'] = master_dataframe['filename'].str.split("/").str[0]

In [ ]:
# Extract the Quarter details from the file path 
master_dataframe['Quarter_details'] =  master_dataframe['filename'].str[5:9]

### 4.2 Extract the Company key

In [ ]:
# Extract the company index code.
master_dataframe['Company_Key'] = master_dataframe['Text_data'].apply(lambda x: extract_company_index(x)) 

In [ ]:
# Removing the special charaters from the code
master_dataframe['Company_Key'] = master_dataframe['Company_Key'].str[6:-2]

In [ ]:
# type casting the key to integer
master_dataframe['Company_Key'] = master_dataframe['Company_Key'].astype(int)

In [ ]:
# showing the top 5 records
master_dataframe.head()

Company_Key                                          Text_data  \
0       755001  b'<Header>\r\n<FileStats>\r\n    <FileName>201...   
1       755001  b'<Header>\r\n<FileStats>\r\n    <FileName>201...   
2       755001  b'<Header>\r\n<FileStats>\r\n    <FileName>201...   
3       755001  b'<Header>\r\n<FileStats>\r\n    <FileName>201...   
4       755001  b'<Header>\r\n<FileStats>\r\n    <FileName>201...   

  Quarter_details  Year                                           filename  \
0            QTR1  2011  2011/QTR1/20110203_10-K_edgar_data_755001_0001...   
1            QTR1  2012  2012/QTR1/20120201_10-K_edgar_data_755001_0001...   
2            QTR1  2013  2013/QTR1/20130130_10-K_edgar_data_755001_0001...   
3            QTR1  2014  2014/QTR1/20140129_10-K_edgar_data_755001_0001...   
4            QTR1  2015  2015/QTR1/20150128_10-K_edgar_data_755001_0001...   

  company_name  cik_filename  actual_list  
0  UNITIL CORP        755001            1  
1  UNITIL CORP        755001            1  
2  UNITIL CORP        755001            1  
3  UNITIL CORP        755001            1  
4  UNITIL CORP        755001            1


#### Note:
For most of the 10-k fillings has the company code in the filename but it is not the case for all the files. SO in the below step we am comparing the company code from the filename to inside the file.


<div class="alert alert-block alert-info">
    <b>FYI:</b> Since there are many file in the zip files it is time consuming and colab is crashing sometimes to process all the files. So we are extracting the the files based on the file name. 
</div> 

In [ ]:
# extracting the company code from the file 
master_dataframe['cik_filename'] = master_dataframe['filename'].str.split('_').str[4]

In [ ]:
# showing the top 5 records
master_dataframe['cik_filename'] = master_dataframe['cik_filename'].astype(int)

In [ ]:
master_dataframe.head(2)

Company_Key                                          Text_data  \
0       755001  b'<Header>\r\n<FileStats>\r\n    <FileName>201...   
1       755001  b'<Header>\r\n<FileStats>\r\n    <FileName>201...   

  Quarter_details  Year                                           filename  \
0            QTR1  2011  2011/QTR1/20110203_10-K_edgar_data_755001_0001...   
1            QTR1  2012  2012/QTR1/20120201_10-K_edgar_data_755001_0001...   

  company_name  cik_filename  
0  UNITIL CORP        755001  
1  UNITIL CORP        755001

In [ ]:
# comparing of the company key from the file is same as the company key from text 
master_dataframe[master_dataframe['Company_Key'] != master_dataframe['cik_filename']]

Company_Key                                          Text_data  \
14         57183  b'<Header>\r\n<FileStats>\r\n    <FileName>201...   
20         20947  b"<Header>\r\n<FileStats>\r\n    <FileName>201...   
21         20947  b'<Header>\r\n<FileStats>\r\n    <FileName>201...   
22         53456  b'<Header>\r\n<FileStats>\r\n    <FileName>201...   
24       1407703  b'<Header>\r\n<FileStats>\r\n    <FileName>201...   
..           ...                                                ...   
537        81100  b'<Header>\r\n<FileStats>\r\n    <FileName>201...   
641        13372  b'<Header>\r\n<FileStats>\r\n    <FileName>201...   
690       754737  b'<Header>\r\n<FileStats>\r\n    <FileName>201...   
692       754737  b'<Header>\r\n<FileStats>\r\n    <FileName>201...   
693       754737  b'<Header>\r\n<FileStats>\r\n    <FileName>201...   

    Quarter_details Year                                           filename  \
14              NaN  NaN  2015/QTR4/20151124_10-K_edgar_data_1126956_000...   
20              NaN  NaN  2011/QTR1/20110216_10-K_edgar_data_1031296_000...   
21              NaN  NaN  2012/QTR1/20120228_10-K_edgar_data_1031296_000...   
22              NaN  NaN  2013/QTR1/20130225_10-K_edgar_data_1031296_000...   
24              NaN  NaN  2015/QTR1/20150217_10-K_edgar_data_1031296_000...   
..              ...  ...                                                ...   
537             NaN  NaN  2016/QTR1/20160229_10-K_edgar_data_1085392_000...   
641             NaN  NaN  2017/QTR1/20170223_10-K_edgar_data_72741_00000...   
690             NaN  NaN  2016/QTR1/20160226_10-K_edgar_data_91882_00007...   
692             NaN  NaN  2018/QTR1/20180223_10-K_edgar_data_91882_00007...   
693             NaN  NaN  2019/QTR1/20190228_10-K_edgar_data_91882_00007...   

                           company_name  cik_filename  
14                       LACLEDE GAS CO       1126956  
20   CLEVELAND ELECTRIC ILLUMINATING CO       1031296  
21   CLEVELAND ELECTRIC ILLUMINATING CO       1031296  
22      JERSEY CENTRAL POWER & LIGHT CO       1031296  
24          FirstEnergy Solutions Corp.       1031296  
..                                  ...           ...  
537              PUGET SOUND ENERGY INC       1085392  
641                   NSTAR ELECTRIC CO         72741  
690                          SCANA CORP         91882  
692                          SCANA CORP         91882  
693                          SCANA CORP         91882  

[79 rows x 7 columns]

- 79 cases there is an mismatch between the company key from the file and company key from text 

In [ ]:
master_dataframe.shape

(888, 7)

### 4.3 Creating an flag column to highlight the mismatch cases 

In [ ]:
#Creating an flag column to highlioght the mismatch cases 
master_dataframe['actual_list'] = np.where(master_dataframe['Company_Key'].isin(company_list['Index_Key'].unique()), 1,0)

In [ ]:
master_dataframe['actual_list'].value_counts()

1    809
0     79
Name: actual_list, dtype: int64

In [ ]:
master_dataframe.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 809 entries, 0 to 887
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Company_Key      809 non-null    int64 
 1   Text_data        809 non-null    object
 2   Quarter_details  809 non-null    object
 3   Year             809 non-null    int64 
 4   filename         809 non-null    object
 5   company_name     809 non-null    object
 6   cik_filename     809 non-null    int64 
 7   actual_list      809 non-null    int64 
dtypes: int64(4), object(4)
memory usage: 56.9+ KB


In [ ]:
master_dataframe.drop(['actual_list'], axis=1, inplace=True)

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4913: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


## 5. Exporting the files to the drive

In [ ]:
# exporting the fikes to the storage system
master_dataframe.to_csv("/content/drive/MyDrive/SPM_files/Selected_10k_v2.csv", index=None)

In [7]:
# code to understan the size of the file 
from sys import getsizeof
print("Size of the dataframe : {} bytes".format(getsizeof(master_dataframe)))

Size of the dataframe : 609765017 bytes


In [ ]:
## End of the Notebook
